### ISSUE TITLE
why no namespace selector in service selector

### ISSUE THREAD

**Issue by npnwdev** [kubernetes/kubernetes#109630]
service uses any backend pods that match labels specified under the selector. 
Since service, pods are both namespaced resources, why service is allowed to select any pods across namespaces?
wouldn't that be better to include a namespace selector to confine the selection to the current namespace or use any namespace (or even all) per user preference?

**Comment by chymy** [comment]
> why service is allowed to select any pods across namespaces? 

There's no such case. If across namespaces, The endpoint corresponding to service is  `<none>`

**Comment by npchandran** [comment]
hey, @chymy Thanks for your response. Unfortunately, I'm not very clear. Could you rephrase, please?

Before that, see if my understanding is correct. Is the service supposed to select any pod within or across namespaces that match the label provided in the selector?

**Comment by chymy** [comment]
service  selects the pod that matches service's  selector  under the same namespace. If there is no such pods in the same namespace, The endpoint corresponding to service is `<none>`.
you can create services without selectors: https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors

**Comment by npnwdev** [comment]
Thanks, @chymy for the clarification. 
could you help me with the spec for "under the same namespace"? 
I am looking for the exact definition that the service selector is applied only on the service's namespace... 
I couldn't find the info at https://kubernetes.io/docs/concepts/services-networking/service/#defining-a-service

**Comment by chymy** [comment]
You can execute the following command:
`kubectl explain service.metadata.namespace`
`kubectl explain service`

**Comment by npchandran** [comment]
hmm, I couldn't find that info yet. a miss in the documentation?

```
user@my-cluster:~$ kubectl explain service.spec.selector
KIND:     Service
VERSION:  v1

FIELD:    selector <map[string]string>

DESCRIPTION:
     Route service traffic to pods with label keys and values matching this
     selector. If empty or not present, the service is assumed to have an
     external process managing its endpoints, which Kubernetes will not modify.
     Only applies to types ClusterIP, NodePort, and LoadBalancer. Ignored if
     type is ExternalName. More info:
     https://kubernetes.io/docs/concepts/services-networking/service/
user@my-cluster:~$ 
```

**Comment by thockin** [comment]
Namespace is an authz subdomain.  Crossing namespaces arbitrarily is a security hole (I can take your app andf put it on the internet without your permission).

While we have started to explore a concept of "reference policy" (part of Gateway API) I don't think we want it for Service.

What problem are you trying to solve?

**Comment by npchandran** [comment]
service.spec.selector selects backend pods for the service.

I was in an assumption that the service.spec.selector can select pods across namespaces as there is no way to specify a namespace selector. 

According to @chymy service.spec.selector is applied only to the namespace where the service is located. That answers my question. 

Now, All I'm looking for is a reference for @chymy statement.
i.e where it's mentioned that the service.spec.selector is applied to the service's namespace?

I looked at explain service.spec.selector (https://github.com/kubernetes/kubernetes/issues/109630#issuecomment-1110014835), but that does not explain whether the selector is applied to the current namespace or across namespaces

From my understanding, the service.spec.selector is applied to the current namespace but that's not explicitly documented.

**Comment by thockin** [comment]
Service selection absolutely happens in a single namespace.  I'd be happy
to look at PRs if you think the comments can be improved :)

**Comment by npchandran** [comment]
Sure. yes, I agree with the behavior. I guess the doc can be made explicit
